server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/service/ExportService.java

```diff
@@ -45,33 +48,39 @@
     private static final Logger LOGGER = LoggerFactory.getLogger(ExportService.class);
     private static final String EXPLANATION = "The backup of your mailboxes has been exported to you";
     private static final String FILE_PREFIX = "mailbox-backup-";
-    public static final boolean LAZY_RESOURCE_CLEANUP = false;
 
     private final MailboxBackup mailboxBackup;
     private final BlobStore blobStore;
     private final BlobExportMechanism blobExport;
     private final UsersRepository usersRepository;
 
-    public ExportService(MailboxBackup mailboxBackup, BlobStore blobStore, BlobExportMechanism blobExport, UsersRepository usersRepository) {
+    @Inject
+    ExportService(MailboxBackup mailboxBackup, BlobStore blobStore, BlobExportMechanism blobExport, UsersRepository usersRepository) {
         this.mailboxBackup = mailboxBackup;
         this.blobStore = blobStore;
         this.blobExport = blobExport;
         this.usersRepository = usersRepository;
     }
 
     public Mono<Task.Result> exportUserMailboxesData(Username username) {
-        return Mono.using(
-            () -> zipMailboxesContent(username),
-            inputStream -> blobStore.save(blobStore.getDefaultBucketName(), inputStream, BlobStore.StoragePolicy.LOW_COST)
+        return Mono.usingWhen(
+            Mono.fromCallable(() -> zipMailboxesContent(username)),
+            inputStream -> saveAndExportThenDeleteBlob(username, inputStream),
+            inputStream -> Mono.fromRunnable(() ->
+                Throwing.consumer(InputStream::close).sneakyThrow())
+        );
+    }
+
+    private Mono<Task.Result> saveAndExportThenDeleteBlob(Username username, InputStream inputStream) {
+        return Mono.defer(() ->
+            blobStore.save(blobStore.getDefaultBucketName(), inputStream, BlobStore.StoragePolicy.LOW_COST)
                 .flatMap(blobId -> export(username, blobId)
-                    .doFinally(any -> deleteBlob(blobId))
-                    .thenReturn(Task.Result.COMPLETED))
-                .onErrorResume(e -> {
-                    LOGGER.error("Error exporting mailboxes of user: {}", username.asString(), e);
-                    return Mono.just(Task.Result.PARTIAL);
-                }),
-            Throwing.consumer(InputStream::close).sneakyThrow(),
-            LAZY_RESOURCE_CLEANUP);
+                    .doFinally(any -> deleteBlob(blobId).block()))
```

Comment:
If you replace the `doFinally` by a `then`, and remove the `block`, I believe it works.

Reply:
then() only run when the Mono upstream complete successfully. In case something bad happens in the export process we still want to clean up the blob right?

Reply:
good point... still convinced we shouldn't need a `block()` there though, but I'm out of ideas ^^'

Reply:
Or maybe using an `await` regarding the test?

Reply:
i'm not a fan of block() either. But @trantienduchn said it was ok to do a block() in the doFinally(), because according to the reator doc doFinally won't be executed until the Mono upstream complete, so there won't be any blocking issue here. 

I think it makes sense but personally, i prefer subscribeOn with a scheduler.

Reply:
I have switched to usingWhen() for this case so we don't need to use block() anymore.